// gatsby-recipes/dist/renderer/render.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _react = _interopRequireWildcard(require("react"));

var _betterQueue = _interopRequireDefault(require("better-queue"));

var _lodash = _interopRequireDefault(require("lodash"));

var _mitt = _interopRequireDefault(require("mitt"));

var _resources = _interopRequireDefault(require("../resources"));

var _reconciler = _interopRequireDefault(require("./reconciler"));

var _errorBoundary = _interopRequireDefault(require("./error-boundary"));

var _transformToPlanStructure = _interopRequireDefault(require("./transform-to-plan-structure"));

var _parentResourceProvider = require("./parent-resource-provider");

var _stepComponent = require("./step-component");

var _inputProvider = require("./input-provider");

var _resourceProvider = require("./resource-provider");

const errorCache = new Map();

const GlobalsContext = _react.default.createContext({});

const useGlobals = () => (0, _react.useContext)(GlobalsContext);

const GlobalsProvider = GlobalsContext.Provider;

const getUserProps = props => {
  // eslint-disable-next-line
  const {
    mdxType,
    children,
    ...userProps
  } = props;
  return userProps;
};

const SetResourcesProvider = _react.default.createContext();

let resourcesCache;

const Wrapper = ({
  children,
  inputs,
  isApply,
  resultCache,
  inFlightCache,
  queue
}) => {
  // eslint-disable-next-line
  const [resourcesList, setResources] = (0, _react.useState)(resourcesCache || []);
  resourcesCache = resourcesList;
  return /*#__PURE__*/_react.default.createElement(_errorBoundary.default, null, /*#__PURE__*/_react.default.createElement(GlobalsProvider, {
    value: {
      mode: isApply ? `apply` : `plan`,
      resultCache,
      inFlightCache,
      queue
    }
  }, /*#__PURE__*/_react.default.createElement(SetResourcesProvider.Provider, {
    value: setResources
  }, /*#__PURE__*/_react.default.createElement(_resourceProvider.ResourceProvider, {
    value: resourcesList
  }, /*#__PURE__*/_react.default.createElement(_inputProvider.InputProvider, {
    value: inputs
  }, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: /*#__PURE__*/_react.default.createElement("p", null, "Loading recipe...")
  }, children))))));
};

const ResourceComponent = ({
  _resourceName: Resource,
  _uuid,
  _type,
  children,
  ...props
}) => {
  const {
    mode,
    resultCache,
    inFlightCache,
    queue
  } = useGlobals();
  const step = (0, _stepComponent.useRecipeStep)();
  const parentResourceContext = (0, _parentResourceProvider.useParentResourceContext)(); // TODO add provider onto context

  const resourceData = handleResource(Resource, { ...parentResourceContext,
    root: process.cwd(),
    _uuid,
    mode,
    resultCache,
    inFlightCache,
    queue
  }, props);
  return /*#__PURE__*/_react.default.createElement(_parentResourceProvider.ParentResourceProvider, {
    data: {
      [Resource]: resourceData
    }
  }, /*#__PURE__*/_react.default.createElement(Resource, null, JSON.stringify({ ...resourceData,
    _props: props,
    _stepMetadata: step,
    _uuid,
    _type
  }), children));
};

const validateResource = (resourceName, context, props) => {
  const userProps = getUserProps(props);

  const {
    error
  } = _resources.default[resourceName].validate(userProps);

  if (error) {
    error.resourceUuid = context._uuid;
  }

  return error;
};

const handleResource = (resourceName, context, props) => {
  // Initialize
  const {
    mode,
    resultCache,
    inFlightCache,
    queue
  } = context; // TODO use session ID to ensure the IDs are unique..

  const trueKey = props._key ? props._key : context._uuid;
  let cacheKey; // Only run apply once per resource

  if (mode === `apply`) {
    cacheKey = mode + ` ` + resourceName + ` ` + trueKey;
  } else {
    cacheKey = JSON.stringify({
      resourceName,
      ...props,
      mode
    });
  } // update global context when results come in.


  const updateResource = result => {
    allResources = allResources.filter(a => a.resourceDefinitions._key);
    const resourceMap = new Map();
    allResources.forEach(r => resourceMap.set(r.resourceDefinitions._key, r));
    const newResource = {
      resourceName,
      resourceDefinitions: props,
      ...result
    };

    if (!_lodash.default.isEqual(newResource, resourceMap.get(trueKey))) {
      resourceMap.set(trueKey, newResource); // TODO Do we need this? It's causing infinite loops
      // setResources([...resourceMap.values()])
    }
  };

  let allResources = (0, _resourceProvider.useResourceContext)();

  if (!errorCache.has(trueKey)) {
    const error = validateResource(resourceName, context, props);
    errorCache.set(trueKey, error);

    if (error) {
      const result = {
        error: `Validation error: ${error.details[0].message}`
      };
      updateResource(result);
      resultCache.set(cacheKey, result);
      return result;
    }
  }

  const cachedResult = resultCache.get(cacheKey);
  const inFlightPromise = inFlightCache.get(cacheKey);

  if (cachedResult) {
    updateResource(cachedResult);
    return cachedResult;
  }

  if (inFlightPromise) {
    throw inFlightPromise;
  }

  const fn = mode === `apply` ? `create` : `plan`;
  let promise;

  try {
    promise = new Promise((resolve, reject) => {
      // Multiple of the same promises can be queued due to re-rendering
      // so this first checks for the cached result again before executing
      // the request.
      const cachedValue = resultCache.get(cacheKey);

      if (cachedValue) {
        resolve(cachedValue);
        updateResource(cachedValue);
      } else {
        _resources.default[resourceName][fn](context, props).then(result => {
          if (fn === `create`) {
            result.isDone = true;
          }

          updateResource(result);
          inFlightCache.set(cacheKey, false);
          return result;
        }).then(result => {
          resultCache.set(cacheKey, result);
          return result;
        }).then(resolve).catch(e => {
          if (e.name === `MissingInfoError`) {
            inFlightCache.delete(cacheKey);
          }

          reject(e);
        });
      }
    });
  } catch (e) {
    throw e;
  }

  inFlightCache.set(cacheKey, promise);
  queue.push(promise);
  throw promise;
};

const render = (recipe, cb, inputs = {}, isApply, isStream, name) => {
  const emitter = (0, _mitt.default)();
  const plan = {};
  const queue = new _betterQueue.default(async (job, cb) => {
    const result = await job;
    cb(null, result);
  }, {
    concurrent: 10000
  });
  const resultCache = new Map();
  const inFlightCache = new Map();
  let result;

  const recipeWithWrapper = /*#__PURE__*/_react.default.createElement(Wrapper, {
    inputs: inputs,
    isApply: isApply,
    resultCache: resultCache,
    inFlightCache: inFlightCache,
    queue: queue
  }, recipe); // Keep calling render until there's remaining resources to render.
  // This let's resources that depend on other resources to pause until one finishes.


  const renderResources = isDrained => {
    result = _reconciler.default.render(recipeWithWrapper, plan, name);
    const resourcesArray = (0, _transformToPlanStructure.default)(result); // Tell UI about updates.

    emitter.emit(`update`, resourcesArray);

    const isDone = () => {
      let result; // Mostly for validation stage that checks that there's no resources
      // in the initial step — this done condition says no resources were found
      // and there's no inflight resource work (resources will be empty until the
      // first resource returns).
      //
      // We use "inFlightCache" because the queue doesn't immediately show up
      // as having things in it.

      if (resourcesArray.length === 0 && ![...inFlightCache.values()].some(a => a)) {
        result = true; // If there's still nothing on the queue and we've drained the queue, that means we're done.
      } else if (isDrained && queue.length === 0) {
        result = true; // If there's one resource & it fails validation, it doesn't go into the queue
        // so we check if inFlightCache is empty & all resources have an error.
      } else if (!resourcesArray.some(r => !r.error) && ![...inFlightCache.values()].some(a => a)) {
        result = true;
      } else {
        result = false;
      }

      return result;
    };

    if (isDone()) {
      // Rerender with the resources and resolve the data from the cache.
      emitter.emit(`done`, resourcesArray);
    }
  };

  const throttledRenderResources = _lodash.default.throttle(renderResources, 100);

  queue.on(`task_finish`, function (taskId, r, stats) {
    throttledRenderResources();
  });
  queue.on(`drain`, () => {
    renderResources(true);
  }); // When there's no resources, renderResources finishes synchronously
  // so wait for the next tick so the emitter listners can be setup first.

  process.nextTick(() => renderResources());

  if (isStream) {
    return emitter;
  } else {
    return new Promise((resolve, reject) => {
      emitter.on(`*`, (type, e) => {
        if (type === `done`) {
          resolve(e);
        }

        if (type === `error`) {
          reject(e);
        }
      });
    });
  }
};

module.exports.render = render;
module.exports.ResourceComponent = ResourceComponent;
